### PR TITLE
Allow host to manually end game 

### PR DESCRIPTION
**Pull request by alankbi**
### List of Changes:

* Create POST endpoint for host to end game before time is up or all players have finished
* Add button on frontend for host to send end game request
* Add relevant test cases for new endpoint and behavior

### Notes:

* There's a small UI bug where after refreshing the page (thus losing host status) the end game button still shows (clicking it will give a permission denied error) - I _think_ this should be resolved once merged with the Redux stuff (which refactors some of the game fetching) but I'm not sure
* At some point in the future if there's a fourth variable to check for game status (right now I think it's gameEnded, allSolved, and timeUp), it might be worth considering refactoring that setup to use a single variable and create a (customizable based on game mode) method/service of sorts that checks when to end the game based on the game mode and stuff sent in

### Screencast and Images:

New end game button

### CODE REVIEW

**Comment by alankbi** [comment]
[Allow host to end game immediately](https://trello.com/c/nJVAIQg7/352-allow-host-to-end-game-immediately)

File: src/main/java/com/codejoust/main/game_object/Game.java
Comment (zpChris) [inline src/main/java/com/codejoust/main/game_object/Game.java:32]: nit, personal preference: "Prematurely" makes it sound like the host did it on accident, so I made a small suggested change to clarify the wording.
```suggestion
    // Boolean to hold whether the host ended the game early
```

File: src/main/java/com/codejoust/main/service/GameManagementService.java
Comment (zpChris) [inline src/main/java/com/codejoust/main/service/GameManagementService.java:284]: Does there have to be any handling on if the game is already ended and this method is called?

Also, I will note that I thought there could arise an issue where the game ends by timer and by manual end game at the same time, but when I tried testing this out by ending the game at around one second, no adverse consequences seemed to occur. 👍

**Review by zpChris — APPROVED** [review]
LGTM! ✅ 

I left a few comments and nits, but nothing at all major. I noted down on another comment for this PR that we may want to wait until the Redux PRs are merged, so we can ensure that we resolve the issue you mentioned where the host still has the "End Game" button even after refreshing.

File: src/main/java/com/codejoust/main/service/GameManagementService.java
Comment (alankbi) [reply src/main/java/com/codejoust/main/service/GameManagementService.java:284]: I don't believe that would cause any issues, since it's just setting that variable to true and canceling the timers - rn I think the one verification we do is if trying to play again when the game hasn't ended yet

However this brings up a good point cause (unless I'm completely missing it) I don't think we do any validation on starting a game or submitting solutions, so that could probably totally screw things up lol. Probably a lower priority since it's an edge case but something to consider

File: frontend/src/views/Game.tsx
Comment (alankbi) [reply frontend/src/views/Game.tsx:420]: I think that makes sense, I like the first option - to make it more like our current shadowy design, how would you feel about something like this? 

Exit game button

**Comment by alankbi** [comment]
> Got it - in that case, thoughts on waiting to merge this until the Redux PRs are merged? I'll still review this PR, but since there's this bug it may be worth it to fix that before merging.

Sounds good - I think #151 is the last one to be approved so once that goes through I'll merge those before updating this one and merging

> I think that makes sense - would the game status variable hold an enum or something that has all the different game status possibilities?

I was thinking if the game object holds an enum for the game mode, then it could just be a boolean for gameOver and whatever service or method that handles the game status could use the game mode/other details to determine when the game ends. Unless there's ever a game mode that has more than just game active and game over, then yeah an enum probably makes sense

File: frontend/src/views/Game.tsx
Comment (zpChris) [reply frontend/src/views/Game.tsx]: That redesign looks fantastic! Much better than my short attempts. 👍

File: src/main/java/com/codejoust/main/service/GameManagementService.java
Comment (zpChris) [reply src/main/java/com/codejoust/main/service/GameManagementService.java:284]: > I don't believe that would cause any issues, since it's just setting that variable to true and canceling the timers - rn I think the one verification we do is if trying to play again when the game hasn't ended yet

That makes sense - if we ever have statistics or show on the UI how the game ended (out of time, ended game early, everybody submitted), then we may need to update this method - otherwise this probably isn't an issue.

Comment (alankbi) [reply src/main/java/com/codejoust/main/service/GameManagementService.java:284]: I think like preventing users from submitting after the game is over, or from starting a game when it's already active, etc.

**Comment by alankbi** [comment]
Just tested locally and the bug should be resolved - I'll merge this tomorrow in case @zpChris you want to re-review at all

**Pull request merged**
